ed_sensor_integration/tools/table_modeler.cpp

```diff
@@ -0,0 +1,161 @@
+    targetToSource = reg.getFinalTransformation().inverse();
+
+    //
+    // Transform target back in source frame
+    pcl::transformPointCloud (*cloud_tgt, *output, targetToSource);
+
+
+
+    //add the source to the transformed target
+    *output += *cloud_src;
+
+    final_transform = targetToSource;
+ }
+
+
+int main(int argc, char **argv) {
+
+    // open pcd files
+
+    if (argc < 2)
+    {
+        std::cout << "Usage:\n\n   table_modeler FILENAME1 FILENAME2 ...\n\n";
+        return 1;
+    }
+
+    //pcl::PointCloud<pcl::PointXYZRGB>::Ptr inputs[argc-1]; //(new pcl::PointCloud<pcl::PointXYZRGB>);
+    std::vector <pcl::PointCloud<pcl::PointXYZRGB>::Ptr, Eigen::aligned_allocator <pcl::PointCloud <pcl::PointXYZRGB>::Ptr > > inputs;
+
+    std::vector<int> indices;
+
+    std::cout << "starting to open" << std::endl;
+
+    for (int i = 1; i < argc; ++i)
+    {
+        std::string name = std::string(argv[i]);
+        pcl::PointCloud<pcl::PointXYZRGB>::Ptr m(new pcl::PointCloud<pcl::PointXYZRGB>);
+        std::cout << "attempting to open " << name << std::endl;
+        if (pcl::io::loadPCDFile<pcl::PointXYZRGB> (name, *m) == -1) //* load the file
+        {
+            //PCL_ERROR ("Couldn't read file" + name + "\n");
+            return -1;
+        }
+        std::cout << "opened " << name << std::endl;
+        pcl::removeNaNFromPointCloud(*m, *m, indices);
+        std::cout << "removed NAN from " << name << std::endl;
+        inputs.push_back(m);
+    }
+
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr result (new pcl::PointCloud<pcl::PointXYZRGB>), source, target;
+    Eigen::Matrix4f GlobalTransform = Eigen::Matrix4f::Identity (), pairTransform;
+
+    for (int i = 1; i < argc-1; ++i)
+    {
+        // read json metadata
+        /*
+        std::ifstream metadata_file(boost::filesystem::change_extension(argv[i-1], ".json").string(), std::ifstream::binary);
+        Json::Value metadata;
+        metadata_file >> metadata;
+        */
+        float x, y, z, qx, qy, qz, qw;
+
+        const float n = 2.0f/(qx*qx+qy*qy+qz*qz+qw*qw);
+        Eigen::Matrix4f Transform = Eigen::Matrix4f::Identity();/* {
+            {1.0f - n*qy*qy - n*qz*qz, n*qx*qy - n*qz*qw, n*qx*qz + n*qy*qw, x},
+            {n*qx*qy + n*qz*qw, 1.0f - n*qx*qx - n*qz*qz, n*qy*qz - n*qx*qw, y},
+            {n*qx*qz - n*qy*qw, n*qy*qz + n*qx*qw, 1.0f - n*qx*qx - n*qy*qy, z},
+            {0.0f, 0.0f, 0.0f, 1.0f}}; */
+
+        Transform(0,0) = 1.0f - n*qy*qy - n*qz*qz;
+        Transform(0,1) = n*qx*qy - n*qz*qw;
+        Transform(0,2) = n*qx*qz + n*qy*qw;
+        Transform(0,3) = x;
+        Transform(1,0) = n*qx*qy + n*qz*qw;
+        Transform(1,1) = 1.0f - n*qx*qx - n*qz*qz;
+        Transform(1,2) = n*qy*qz - n*qx*qw;
+        Transform(1,3) = y;
+        Transform(2,0) = n*qx*qz - n*qy*qw;
+        Transform(2,1) = n*qy*qz + n*qx*qw;
+        Transform(2,2) = 1.0f - n*qx*qx - n*qy*qy;
+        Transform(2,3) = z;
```

Comment:
No idea what we are doing here. But it looks like some quaternion stuff that probably is already in some library.